site/content/blog/2019-10-25-instrumenting-go-code-via-ast.md

```diff
@@ -0,0 +1,315 @@
+
+To simplify this, we've added a simple tracing module:
+```go
+package tracing
+
+import (
+	opentracing "github.com/opentracing/opentracing-go"
+	"github.com/uber/jaeger-lib/metrics"
+
+	"github.com/uber/jaeger-client-go"
+	jaegercfg "github.com/uber/jaeger-client-go/config"
+	jaegerlog "github.com/uber/jaeger-client-go/log"
+	"context"
+)
+
+var initialized = false
+
+func Initialize() error{
+	cfg := jaegercfg.Configuration{
+		Sampler: &jaegercfg.SamplerConfig{
+			Type:  jaeger.SamplerTypeConst,
+			Param: 1,
+		},
+		Reporter: &jaegercfg.ReporterConfig{
+			LogSpans: true,
+		},
+	}
+
+	_, err := cfg.InitGlobalTracer(
+		"mattermost",
+		jaegercfg.Logger(jaegerlog.StdLogger),
+		jaegercfg.Metrics( metrics.NullFactory),
+	)
+	if err != nil {
+		return err
+	}
+
+	initialized = true
+
+	return nil
+}
+
+func StartRootSpanByContext(ctx context.Context, operationName string)(opentracing.Span, context.Context) {
+	return opentracing.StartSpanFromContext(ctx, operationName)
+}
+
+func StartSpanWithParentByContext(ctx context.Context, operationName string) (opentracing.Span, context.Context) {
+	parentSpan := opentracing.SpanFromContext(ctx)
+
+	if parentSpan == nil {
+		return StartRootSpanByContext(ctx, operationName)
+	}
+
+	return opentracing.StartSpanFromContext(ctx, operationName, opentracing.ChildOf(parentSpan.Context()))
+}
+```
+
+
+Now, wrapping an API call in a span is very easy:
+
+```go
+
+func apiCall(c *Context, w http.ResponseWriter, r *http.Request) {
+	span, ctx := tracing.StartSpanWithParentByContext(c.App.Context, "api4:apiCall")
+	c.App.Context = ctx
+    defer span.Finish()
+
+    // perform actual request handling
+}
+```
+
+Now, each time the `apiCall` handler is invoked, it will be measured and traced on Jaeger.
+
+The problem is - we have a rather large API surface (around 300+ handlers) and it would be a really tough task to instrument all the handlers by hand. That's where our story begins.
+
+
+## Go AST
+
+So what is AST really? Well, to quote Wikipedia:
+
+> In computer science, an abstract syntax tree (AST), or just syntax tree, is a tree representation of the abstract syntactic structure of source code written in a programming language. Each node of the tree denotes a construct occurring in the source code.
+
+![ast example](https://upload.wikimedia.org/wikipedia/commons/thumb/c/c7/Abstract_syntax_tree_for_Euclidean_algorithm.svg/1024px-Abstract_syntax_tree_for_Euclidean_algorithm.svg.png?1572188936106)
+
+Basically, AST is a tree like representation of your source code.
+
+Why do we need it?
+
+Parsing AST allows us to refactor code on a statement level, meaning, instead of finding methods and fields using string search, we can find them by their actual structure. This gives us incredible flexibility in writing custom go refactoring tools.
```

Comment:
```suggestion
Parsing the code into an AST allows us to refactor code on a statement level, meaning, instead of finding methods and fields using string search, we can find them by their actual structure. This gives us incredible flexibility in writing custom go refactoring tools.
```